### PR TITLE
Support backend storage driver: Aliyun OSS. 

### PR DESCRIPTION
**Pull request by soulseen**
Signed-off-by: xiaoyang zhu <zhuxiaoyang1996@gmail.com>

fix https://github.com/goharbor/harbor-operator/issues/638

### CODE REVIEW

File: controllers/goharbor/registry/deployments.go [hunk 1]
```diff
@@ -191,6 +191,23 @@ func (r *Reconciler) GetDeployment(ctx context.Context, registry *goharborv1.Reg
 		})
 	}
 
+	if registry.Spec.Storage.Driver.AliOSS != nil && registry.Spec.Storage.Driver.AliOSS.AccessSecretRef != "" {
+		volumes = append(volumes, corev1.Volume{
+			Name: "gcs-key",
```
Comment (bitsf) [inline controllers/goharbor/registry/deployments.go:196]: this name is not correct

File: controllers/goharbor/registry/deployments.go [hunk 2]
```diff
@@ -559,6 +576,24 @@ func (r *Reconciler) ApplyGcsStorageEnvs(ctx context.Context, registry *goharbor
 	return nil
 }
 
+func (r *Reconciler) ApplyAliOSSStorageEnvs(ctx context.Context, registry *goharborv1.Registry, deploy *appsv1.Deployment) error {
+	regContainer := &deploy.Spec.Template.Spec.Containers[registryContainerIndex]
+
+	regContainer.Env = append(regContainer.Env, corev1.EnvVar{
+		Name: "REGISTRY_STORAGE_OSS_ACCESSKEYSECRET",
```
Comment (bitsf) [inline controllers/goharbor/registry/deployments.go:583]: REGISTRY_STORAGE_ALIOSS_ACCESSKEYSECRET  ?

File: apis/goharbor.io/v1beta1/harborcluster_types.go [hunk 3]
```diff
@@ -11,6 +11,7 @@ import (
 const (
 	KindDatabaseZlandoPostgreSQL = "Zlando/PostgreSQL"
 	KindDatabasePostgreSQL       = "PostgreSQL"
+	KindStorageAliOSS            = "aliOSS"
```
Comment (yxxhero) [inline apis/goharbor.io/v1beta1/harborcluster_types.go:14]: aliOSS --> AliOSS

File: controllers/goharbor/registry/deployments.go (hunk 2 above)
Comment (soulseen) [reply controllers/goharbor/registry/deployments.go:583]: I think it should be consistent with helm.
https://github.com/goharbor/harbor-helm/blob/master/templates/registry/registry-secret.yaml#L33

File: apis/goharbor.io/v1beta1/harborcluster_types.go [hunk 4]
```diff
@@ -216,7 +217,7 @@ type ZlandoPostgreSQLSpec struct {
 
 type Storage struct {
 	// Kind of which storage service to be used. Only support MinIO now.
-	// +kubebuilder:validation:Enum={MinIO,S3,Swift,FileSystem,Azure,Gcs}
+	// +kubebuilder:validation:Enum={MinIO,S3,Swift,FileSystem,Azure,Gcs, AliOSS}
```
Comment (cndoit18) [inline apis/goharbor.io/v1beta1/harborcluster_types.go:220]:
```suggestion
	// +kubebuilder:validation:Enum={MinIO,S3,Swift,FileSystem,Azure,Gcs,AliOSS}
```

File: apis/goharbor.io/v1beta1/chartmuseum_types.go [hunk 5]
```diff
@@ -161,6 +161,23 @@ type ChartMuseumChartStorageDriverSpec struct {
...
+
+	// +kubebuilder:validation:Optional
+	AccessKeyID string `json:"accessKeyID,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	AccessSecretRef string `json:"accessSecretRef,omitempty"`
+
+	// +kubebuilder:validation:Required
+	BucketName string `json:"bucketName"`
 }
```
Comment (bitsf) [inline apis/goharbor.io/v1beta1/chartmuseum_types.go:181]: should add PathPrefix

**Comment by bitsf** [comment]
can't install harbor successful
image

**Review by bitsf — COMMENTED** [review]
refer pkg/cluster/controllers/harbor/harbor.go:158

**Review by bitsf — COMMENTED** [review]
refer controllers/goharbor/harbor/storage.go:108

**Review by bitsf — COMMENTED** [review]
Access id not inject

image
image

**Comment by soulseen** [comment]
@bitsf about Access id not inject, may I bother you for sharing your Ali OSS secret yaml?

**Comment by bitsf** [comment]
close as #991

**Pull request closed without merging**